app/models/spree/payment/processing.rb

```diff
@@ -4,19 +4,20 @@ module Spree
   class Payment < ActiveRecord::Base
     module Processing
       def process!
-        return unless payment_method&.source_required?
+        return unless ready_to_process?
 
-        raise Core::GatewayError, Spree.t(:payment_processing_failed) unless source
-
-        return if processing?
-
-        unless payment_method.supports?(source)
-          invalidate!
-          raise Core::GatewayError, Spree.t(:payment_method_not_supported)
+        if payment_method.auto_capture?
+          purchase!
+        else
+          authorize!
         end
+      end
+
+      def process_offline!
+        return unless ready_to_process?
 
         if payment_method.auto_capture?
-          purchase!
+          charge_offline!
```

Comment:
Like I suggested in `#process_payments_offline!` we could remove this duplication pretending we follow the [strategy pattern](https://refactoring.guru/design-patterns/strategy/ruby/example) :joy: To me, making it clear we follow it it's more a signal for further refactorings of this module, which should be extracted into services IMO.

Reply:
Going to resolve this one and let's just keep any discussion in the one above 👍